File: Sistema Distribuidos/servidores/servidor2_grpc.py
import grpc
import os
import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from concurrent import futures
import mensagem_pb2
import mensagem_pb2_grpc
from comum.lamport import LamportClock
from comum.relogio import RelogioFisico
import time

# Função simples de log para testes
def log_event(processo, evento, ts_logico, ts_fisico):
    with open(f"{processo}_log.txt", "a") as f:
        f.write(f"[{ts_fisico:.2f} | {ts_logico}] {evento}\n")


class MensagemServiceImpl(mensagem_pb2_grpc.MensagemServiceServicer):
    def __init__(self):
        self.clock_logico = LamportClock()
        self.clock_fisico = RelogioFisico()

    def EnviarMensagem(self, request, context):
        ts_logico = self.clock_logico.update(request.timestamp)
        ts_fisico = self.clock_fisico.agora()

        evento = f"Mensagem recebida de {request.de} para {request.para}: {request.conteudo}"
        log_event("servidor", evento, ts_logico, ts_fisico)

        return mensagem_pb2.Confirmacao(recebido=True)

    def ChatMensagens(self, request_iterator, context):
        for mensagem in request_iterator:
            ts_logico = self.clock_logico.update(mensagem.timestamp)
            ts_fisico = self.clock_fisico.agora()

            evento = f"[Chat] {mensagem.de} para {mensagem.para}: {mensagem.conteudo}"
            log_event("servidor", evento, ts_logico, ts_fisico)

            yield mensagem_pb2.Mensagem(
                de=mensagem.para,
                para=mensagem.de,
                conteudo=f"Eco: {mensagem.conteudo}",
                timestamp=ts_logico
            )

    def ObterTempoVivo(self, request, context):
        ts_fisico = self.clock_fisico.agora()
        return mensagem_pb2.Tempo(tempo=ts_fisico)

    def AjustarTempo(self, request, context):
        self.clock_fisico.ajustar(request.delta)
        return mensagem_pb2.Vazio()  # Ajustado para retornar Vazio conforme proto


def serve():
    porta = os.getenv("PORTA_SERVIDOR", "50052")
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    mensagem_pb2_grpc.add_MensagemServiceServicer_to_server(MensagemServiceImpl(), server)
    server.add_insecure_port(f"[::]:{porta}")
    print(f"Servidor gRPC ativo em porta {porta}...")
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    serve()

File: Sistema Distribuidos/servidores/servidor2_grpc.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const fs = require('fs');

const PROTO_PATH = './mensagem.proto';
const packageDefinition = protoLoader.loadSync(PROTO_PATH);
const proto = grpc.loadPackageDefinition(packageDefinition).mensagens;

let clockLogico = 0;
let deltaAjuste = 0;

function agora() {
  return Date.now() / 1000 + deltaAjuste;
}

function logEvent(processo, evento, tsLogico, tsFisico) {
  fs.appendFileSync(`${processo}_log.txt`, `[${tsFisico.toFixed(2)} | ${tsLogico}] ${evento}\n`);
}

function EnviarMensagem(call, callback) {
  clockLogico = Math.max(clockLogico, call.request.timestamp) + 1;
  const tsFisico = agora();
  logEvent("servidor", `Mensagem recebida de ${call.request.de} para ${call.request.para}: ${call.request.conteudo}`, clockLogico, tsFisico);
  callback(null, { recebido: true });
}

function ChatMensagens(call) {
  call.on('data', (mensagem) => {
    clockLogico = Math.max(clockLogico, mensagem.timestamp) + 1;
    const tsFisico = agora();

    // Log no servidor
    logEvent("servidor", `[Chat] ${mensagem.de} para ${mensagem.para}: ${mensagem.conteudo}`, clockLogico, tsFisico);

    // Log no remetente
    logEvent(mensagem.de, `Enviou (privado): ${mensagem.conteudo}`, clockLogico, tsFisico);

    // Log no destinatário
    logEvent(mensagem.para, `Recebeu (privado) de ${mensagem.de}: ${mensagem.conteudo}`, clockLogico, tsFisico);

    call.write({
      de: mensagem.para,
      para: mensagem.de,
      conteudo: `Eco: ${mensagem.conteudo}`,
      timestamp: clockLogico
    });
  });

  call.on('end', () => {
    call.end();
  });
}


function ObterTempoVivo(call, callback) {
  callback(null, { tempo: agora() });
}

function AjustarTempo(call, callback) {
  deltaAjuste += call.request.delta;
  callback(null, {}); // Retorna mensagem Vazio
}

function main() {
  const servidor = new grpc.Server();
  servidor.addService(proto.MensagemService.service, {
    EnviarMensagem,
    ChatMensagens,
    ObterTempoVivo,
    AjustarTempo
  });

  const porta = process.env.PORTA_SERVIDOR || '50052';
  servidor.bindAsync(`0.0.0.0:${porta}`, grpc.ServerCredentials.createInsecure(), () => {
    console.log(`Servidor gRPC em JavaScript rodando na porta ${porta}`);
    servidor.start();
  });
}

main();
